salt/_pillar/metalk8s.py

```diff
@@ -49,12 +51,42 @@ def _load_ca(config_data):
     }
 
 
+def _get_endpoints(config="/etc/kubernetes/admin.conf",
+                   context="kubernetes-admin@kubernetes",
+                   namespace="kube-system"):
+    """Find out services endpoints."""
+    endpoints = {}
+
+    if 'kubernetes.show_endpoint' not in __salt__ \
+            or not os.path.exists(config):
+        return endpoints
+
+    for service in ('salt-master', 'registry', 'package-repositories'):
+        try:
+            endpoint = __salt__['kubernetes.show_endpoint'](
+                name=service,
+                namespace=namespace,
+                kubeconfig=config,
+                context=context
+            )
+            endpoint_ip = endpoint["subsets"][0]["addresses"][0]['ip']
```

Comment:
Why throw away port information?

Reply:
No specific reason, I just thought that we'll only use ip in our use case, but it makes sense to add port info as well. BTW, we'll have something like this in the `endpoint["subsets"]` dict:
```
>>> pprint(coreV1.read_namespaced_endpoints('salt-master', 'kube-system').to_dict()['subsets'])
[{'addresses': [{'hostname': None,
                 'ip': '172.21.254.3',
                 'node_name': 'bootstrap',
                 'target_ref': {'api_version': None,
                                'field_path': None,
                                'kind': 'Pod',
                                'name': 'salt-master-bootstrap',
                                'namespace': 'kube-system',
                                'resource_version': '577',
                                'uid': 'f857e39e-5d26-11e9-8308-5254008481d5'}}],
  'not_ready_addresses': None,
  'ports': [{'name': 'requestserver', 'port': 4506, 'protocol': 'TCP'},
            {'name': 'publisher', 'port': 4505, 'protocol': 'TCP'},
            {'name': 'api', 'port': 4507, 'protocol': 'TCP'}]}]
```
Which form do you recommend to return the pillar? Return the dict as it is or make some processing to simplify it (in the pillar I only take the first adress in the subset, but I wonder if it is a good idea)

Reply:
Taking 'the first one' is fine, for now there will be only one anyway, since the bootstrap node is not HA right now anyway.

I think what you may want to return is something like the following:

```json
[
  {
    "node_name": "bootstrap",
    "hostname": null,
    "ip": "172.21.254.3",
    "ports": {
      "requestserver": 4506,
      "publisher": 4505,
      "api": 4507
    }
  }
]
```

Then, in the *ext_pillar*, we set the pillar value to the 'first' element of what this procedure returns, and then consumer of that pillar value can use something like

```jinja
foo: {{ salt_master.ip }}:{{ salt_master.ports.publisher }}
```

or something along those lines.